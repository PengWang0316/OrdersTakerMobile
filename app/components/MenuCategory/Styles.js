import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  sectionHeader: {
    fontWeight: 'bold',
    fontSize: 25,
    marginBottom: 10,
    marginTop: 10,
  },
  itemView: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 10,
    marginBottom: 10,
  },
});
