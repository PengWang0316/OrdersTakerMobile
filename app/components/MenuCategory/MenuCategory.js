import React, { Component } from 'react';
import { connect } from 'react-redux';
import { SectionList, Text, View } from 'react-native';

import { fetchAllMenu } from '../../actions/MenuActions';
import MenuItem from '../MenuItem';
import Styles from './Styles';


type Props = {
  fetchAllMenu: Function,
  containerStyles?: Object,
  numColumns?: number,
  menus: Object,
};

/**
 * The Menu Category component that include all menu item.
 */
export class MenuCategory extends Component<Props> {
  static defaultProps = {
    containerStyles: {},
    numColumns: 2,
  };

  /**
   * Calling the fetchAllMenu to fetch and save menu information.
   * @param {object} props contains prop values.
   */
  constructor(props) {
    super(props);
    if (!props.menus) props.fetchAllMenu();
  }

  /**
   * Putting several item to a view based on the numColums props.
   * @param {object} all param will be supplied from SectionList renderItem method.
   * @return {View} Return a View component with flex: 1 and flexDirection: 'row'.
   */
  _renderItem = ({ index, section }) => {
    const { numColumns } = this.props;
    // % does not equal 0 means this item has already been added by the previous call.
    if (index % numColumns !== 0) return null;
    const itemsArr = [];
    for (let i = index; i < index + numColumns; i++) {
      if (i >= section.data.length) break; // i could over the max number when it reach the end.
      itemsArr.push(<MenuItem key={i} item={section.data[i]} />); // Push a MenuItem component to the array.
    }
    return <View key={index} style={Styles.itemView}>{itemsArr}</View>;
  };

  /**
   * The render method
   * @return {jsx} Return jsx for the component.
   */
  render() {
    const { containerStyles, menus } = this.props;
    return (
      <View style={containerStyles}>
        {menus && (
          <SectionList
            renderItem={params => this._renderItem(params)}
            renderSectionHeader={({ section: { title } }) => (
              <Text style={Styles.sectionHeader}>{title}</Text>
            )}
            sections={menus.map(menu => ({ title: menu.category, data: menu.items }))}
            keyExtractor={(item, index) => item + index}
            // contentContainerStyle={{ flexWrap: 'wrap', flexGrow: 1, flexDirection: 'row' }}
          />
        )}
      </View>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = state => ({
  menus: state.menus,
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchAllMenu: () => dispatch(fetchAllMenu()),
});
export default connect(mapStateToProps, mapDispatchToProps)(MenuCategory);
