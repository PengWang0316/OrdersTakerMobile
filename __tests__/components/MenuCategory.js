import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { MenuCategory } from '../../app/components/MenuCategory/MenuCategory';

jest.mock('../../app/components/MenuItem', () => 'MenuItem');

describe('MenuCategory', () => {
  const defaultProps = {
    fetchAllMenu: jest.fn(),
    containerStyles: {},
    numColumns: 2,
    menus: [{ category: 'category', items: [] }],
  };
  const getShallowComponent = (props = defaultProps) => shallow(<MenuCategory {...props} />);

  test('constructor has menus', () => {
    getShallowComponent();
    expect(defaultProps.fetchAllMenu).not.toHaveBeenCalled();
  });

  test('constructor does not have menus', () => {
    getShallowComponent({ ...defaultProps, menus: null });
    expect(defaultProps.fetchAllMenu).toHaveBeenCalledTimes(1);
  });

  test('_renderItem index 0', () => {
    const component = getShallowComponent();
    const section = {
      data: [
        'data0', 'data1', 'data3',
      ],
    };
    const returnComponent = component.instance()._renderItem({ index: 0, section });

    expect(returnComponent.key).toBe('0');
    expect(returnComponent.props.children.length).toBe(2);
  });

  test('_renderItem index 1', () => {
    const component = getShallowComponent();
    const section = {
      data: [
        'data0', 'data1', 'data3',
      ],
    };
    const returnComponent = component.instance()._renderItem({ index: 1, section });

    expect(returnComponent).toBeNull();
  });

  test('_renderItem index 2', () => {
    const component = getShallowComponent();
    const section = {
      data: [
        'data0', 'data1', 'data3',
      ],
    };
    const returnComponent = component.instance()._renderItem({ index: 2, section });

    expect(returnComponent.key).toBe('2');
    expect(returnComponent.props.children.length).toBe(1);
  });

  test('Snapshot with menus', () => expect(renderer.create(<MenuCategory {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot without menus', () => expect(renderer.create(<MenuCategory {...{ ...defaultProps, menus: null }} />).toJSON()).toMatchSnapshot());
});
