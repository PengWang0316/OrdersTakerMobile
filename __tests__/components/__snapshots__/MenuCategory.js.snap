// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuCategory Snapshot with menus 1`] = `
<View
  style={Object {}}
>
  <RCTScrollView
    data={
      Array [
        Object {
          "data": Array [],
          "title": "category",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    renderSectionHeader={[Function]}
    scrollEventThrottle={50}
    sections={
      Array [
        Object {
          "data": Array [],
          "title": "category",
        },
      ]
    }
    stickyHeaderIndices={
      Array [
        0,
      ]
    }
    stickySectionHeadersEnabled={true}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          style={
            Object {
              "fontSize": 25,
              "fontWeight": "bold",
              "marginBottom": 10,
              "marginTop": 10,
            }
          }
        >
          category
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`MenuCategory Snapshot without menus 1`] = `
<View
  style={Object {}}
/>
`;
